### ISSUE TITLE
collaboration without (user-facing) git

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaComputing/JuliaHub#131]
Collaboration on projects among teams that don't want to use git came up on the JuliaHub call today. I wanted to open an issue to write down some thoughts on it. If two people have done a series of changes and we want to merge those changes, the absolute best way to do it is to use the git-imerge tool:

https://github.com/mhagger/git-imerge

The UI is complicated and not something we should expose to users, but the basic we can use imerge under the hood and it will let us present users with the absolutely minimal conflicts and better still, allows generating normal merges or rebased while only requiring the user to resolve each minimal conflict once. When I have a nasty merge to deal with, this it *the* tool I use. [This video](https://www.youtube.com/watch?v=FMZ2_-Ny_zc) is a great explanation of how imerge works.

So the concept is that we record on the behalf of the user as sequence of implied commits and then we can use imerge to merge different users' branches, presenting them the essential conflicts to resolve in some nice web-based conflict-resolution UI. The quality of this experience seems like it will depend heavily on the quality of the commit sequence that we infer — and fine-grained but sensible commits help keep the essential conflicts smaller. My thought is maybe implicitly committing when the user submits a job? Although that doesn't work if they're developing purely locally and never submitting jobs.

We may also want to simplify the commit history because it's pretty common to make a bunch of changes and then undo them and we would want to drop the intermediate states. A good heuristic would be that if there's a sequence of versions A, B, C and `diff_size(A, C)` is a smaller than `min(diff_size(A, B), diff_size(B, C))` then we would omit B from the version sequence that we consider since B -> C seems to be undoing much of what A -> B did in the first place.

Anyway, this is a complex feature, but the alternative seems to be collaborative editing, which is literally the same problem but in real time.

**Comment by c42f** [comment]
Somehow I've not seen it before, but the concept behind `imerge` looks amazing.

> My thought is maybe implicitly committing when the user submits a job? Although that doesn't work if they're developing purely locally and never submitting jobs.

This part is certainly hard. I think we need to unify the "run a batch job" with "run this code locally" by having a well defined Julia entry point for user code which will make code and data versioning automatic, both locally and remotely.

For the data part, this is basically what the `@datafunc` macro is about in DataSets.jl: `@datafunc` is meant to provide the entry point to the user's code which knows about input and output datasets. (Furthermore, to make the user provide enough syntax that we can infer the possible type of input/output data using only syntax analysis — this will let us provide a richer UI for kicking off jobs.)

When running locally (in the REPL, say), `@datarun` is the equivalent of kicking off a job, and I guess code versioning could be hooked in here as well.

I should say that there's quite a bit of work left to make all this run smoothly. In particular, to refine the API for output datasets, but also to find and fix any design bugs in `@datafunc` and `@datarun`.

**Comment by StefanKarpinski** [comment]
> Somehow I've not seen it before, but the conceptual model behind `imerge` looks amazing.

Yeah, it's a really brilliant approach/tool that is somehow not widely known.

**Comment by StefanKarpinski** [comment]
So a general outline of the approach might be:

1. Record a sequence of candidate commits for each user (snapshots of the project)
2. Prune each user's commit sequence — basically an optimizations problem where we want to choose a subsequence of commits that minimize some loss function on the set of diff sizes. We might do this based on two users' histories: if we can merge later states without conflicts but some earlier intermediate state would have caused conflicts, then we can just drop the earlier snapshots.
3. Run `imerge` on the resulting commit sequences to combine changes from two users. Present the user doing the merge with each essential conflict to resolve once. If we're recording the history ourselves and not exposing it to the end-user, we might want to keep the entire merge grid that imerge generates.

If we had some way to test if some code actually works or not, we could potentially try probably conflict resolutions and see if they work and just do merges automatically if they work. We could do that work speculatively "in the background" and only present users with merge conflicts when there's an actual conflict with what someone else did in the mean time.

**Comment by c42f** [comment]
> Prune each user's commit sequence — basically an optimizations problem where we want to choose a subsequence of commits that minimize some loss function on the set of diff sizes. We might do this based on two users' histories: if we can merge later states without conflicts but some earlier intermediate state would have caused conflicts, then we can just drop the earlier snapshots.

Sounds good. Pruning would be subject to the constraint that code versions which have had side effects (eg, have been used to run a job) must remain in the history in some way. (Though I guess this doesn't quite imply they need to be part of the `imerge` procedure.

**Comment by pfitzseb** [comment]
> Though I guess this doesn't quite imply they need to be part of the imerge procedure.

I guess the question is whether we want to end up with a linear history or are fined with detached commits for jobs, no?

**Comment by StefanKarpinski** [comment]
It's also unclear to me whether surreptitiously doing git merges behind the user's back is the best approach. Another one could be having people "check out" files in order to work on them to avoid conflicts. I know that feels really old fashioned but might be clearer for end users who don't want to use git? But a big part of me thinks that the objection to git is not having to resolve a merge conflict every now and then, but rather that the command-line tool is confusing and if we can just present them with the occasional merge conflict to fix when necessary, things will mostly be fine.

**Comment by pfitzseb** [comment]
FWIW, no one would *need* to use the CLI; VSCode's git tooling is fairly decent (although I have to admit I mainly use it for conflict resolution).
